fastly/pool.go

```diff
@@ -0,0 +1,293 @@
+	PoolTypeHash PoolType = "hash"
+
+	// PoolTypeClient ins a pool that does client direction.
+	PoolTypeClient = "client"
+)
+
+// PoolType is a type of pool.
+type PoolType string
+
+// Pool represents a pool response from the Fastly API.
+type Pool struct {
+	ServiceID string `mapstructure:"service_id"`
+	Version   int    `mapstructure:"version"`
+
+	ID               string   `mapstructure:"id"`
+	Name             string   `mapstructure:"name"`
+	Comment          string   `mapstructure:"comment"`
+	Shield           string   `mapstructure:"shield"`
+	RequestCondition string   `mapstructure:"request_condition"`
+	MaxConnDefault   uint     `mapstructure:"max_conn_default"`
+	ConnectTimeout   uint     `mapstructure:"connect_timeout"`
+	FirstByteTimeout uint     `mapstructure:"first_byte_timeout"`
+	Quorum           uint     `mapstructure:"quorum"`
+	UseTLS           bool     `mapstructure:"use_tls"`
+	TLSCACert        string   `mapstructure:"tls_ca_cert"`
+	TLSCiphers       string   `mapstructure:"tls_ciphers"`
+	TLSClientKey     string   `mapstructure:"tls_client_key"`
+	TLSClientCert    string   `mapstructure:"tls_client_cert"`
+	TLSSNIHostname   string   `mapstructure:"tls_sni_hostname"`
+	TLSCheckCert     bool     `mapstructure:"tls_check_cert"`
+	TLSCertHostname  string   `mapstructure:"tls_cert_hostname"`
+	MinTLSVersion    string   `mapstructure:"min_tls_version"`
+	MaxTLSVersion    string   `mapstructure:"max_tls_version"`
+	Healthcheck      string   `mapstructure:"healthcheck"`
+	Type             PoolType `mapstructure:"type"`
+	OverrideHost     string   `mapstructure:"override_host"`
+}
+
+// poolsByName is a sortable list of pools.
+type poolsByName []*Pool
+
+// Len, Swap, and Less implement the sortable interface.
+func (s poolsByName) Len() int      { return len(s) }
+func (s poolsByName) Swap(i, j int) { s[i], s[j] = s[j], s[i] }
+func (s poolsByName) Less(i, j int) bool {
+	return s[i].Name < s[j].Name
+}
+
+// ListPoolsInput is used as input to the ListPools function.
+type ListPoolsInput struct {
+	// Service is the ID of the service (required).
+	Service string
+
+	// Version is the specific configuration version (required).
+	Version int
+}
+
+// ListPools lists all pools for a particular service and version.
+func (c *Client) ListPools(i *ListPoolsInput) ([]*Pool, error) {
+	if i.Service == "" {
+		return nil, ErrMissingService
+	}
+
+	if i.Version == 0 {
+		return nil, ErrMissingVersion
+	}
+
+	path := fmt.Sprintf("/service/%s/version/%d/pool", i.Service, i.Version)
+	resp, err := c.Get(path, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	var ps []*Pool
+	if err := decodeJSON(&ps, resp.Body); err != nil {
+		return nil, err
+	}
+	sort.Stable(poolsByName(ps))
+	return ps, nil
+}
+
+// CreatePoolInput is used as input to the CreatePool function.
+type CreatePoolInput struct {
+	// Service is the ID of the service. Version is the specific configuration
+	// version. Both fields are required.
+	Service string
```

Comment:
We have a lot of inconsistency with `Service` vs `ServiceID` in the client (which we hope to address this year), and it seems you also have inconsistency in this file (`Pool` struct on line 25 uses `ServiceID`). Therefore, can you please change this to `ServiceID`.

(Similar comments elided for the rest of the review)

Reply:
I tried to keep consistent usage as the rest of the library, the pattern that stood out to me was the use of `ServiceID` in the API response structs for the resources and `Service` in the related `*Input` structs. For example:

* [`ACLEntry`](https://github.com/fastly/go-fastly/blob/e18f1a520c404b08eb1d2c219b9010de494aab47/fastly/acl_entry.go#L10)
    - [`ListACLEntriesInput`](https://github.com/fastly/go-fastly/blob/e18f1a520c404b08eb1d2c219b9010de494aab47/fastly/acl_entry.go#L35)
    - [`GetACLEntryInput`](https://github.com/fastly/go-fastly/blob/e18f1a520c404b08eb1d2c219b9010de494aab47/fastly/acl_entry.go#L68)

* [`Director`](https://github.com/fastly/go-fastly/blob/e18f1a520c404b08eb1d2c219b9010de494aab47/fastly/director.go#L29)
   - [`CreateDirectorInput`](https://github.com/fastly/go-fastly/blob/e18f1a520c404b08eb1d2c219b9010de494aab47/fastly/director.go#L91)
   - [`DeleteDirectorInput`](https://github.com/fastly/go-fastly/blob/e18f1a520c404b08eb1d2c219b9010de494aab47/fastly/director.go#L214)

A quick set of grep heuristics seems to verify this pattern:

    $  grep -ER '\s+ServiceID\s+string.*' . | wc -l
        36
    $ grep -ER '\s+Service\s+string.*' . | wc -l
        180

Given there are 5 operations per resource (list, get, create, update, delete), this works out surprisingly well.

If you're planning on moving the CRUD operation inputs to use a `ServiceID` field and want to start with these resources I'll happily make the changes. Let me know what you think.

Reply:
Ok I'm happy to leave as is and we'll do the cleanup with the wider refactor later in the year 👍